File: databases/src/test/java/victor/training/reactive/databases/MyTest.java
package victor.training.reactive.databases;

import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.ReactiveTransactionManager;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.transaction.reactive.TransactionalOperator;
import reactor.core.publisher.Mono;
import victor.training.reactive.databases.r2dbc.User;
import victor.training.reactive.databases.r2dbc.UserRepository;

import static org.assertj.core.api.Assertions.assertThat;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
@SpringBootTest
public class MyTest {
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private TransactionalOperator operator;
    @Autowired
    private ReactiveTransactionManager rwriteTransactionManager;

    @Test
    @Order(1)
    void explore() {
        assertThat(userRepository.count().block()).isEqualTo(0);
        TransactionalOperator.create(rwriteTransactionManager)
                .execute(status -> {
                    status.setRollbackOnly();
                    return userRepository.save(new User("Jhon"));
                }).blockFirst();

    }

    @Order(2)
    @Test
    void explore2() {
        assertThat(userRepository.count().block()).isEqualTo(0);
        userRepository.save(new User("Jhon")).block();
    }
}

File: spring-webflux/src/test/java/victor/training/reactive/viorel/MyTest.kt
package victor.training.reactive.viorel

import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.repository.ReactiveMongoRepository
import reactor.core.publisher.Mono
import java.util.*

class MyTest {
    @Test
    fun `my test`() {
        val mockRepo = mockk<MyRepo>()
        val myService = MyService(mockRepo)
        every { mockRepo.save(any()) } answers {Mono.just(firstArg())}

        Assertions.assertThat(myService.prod("Rara").block()).isEqualTo("Rara")
    }
}

class MyService(private val myRepo: MyRepo) {
    fun prod(name:String): Mono<String> {
        val doc = MyDoc(name=name)
        return myRepo.save(doc)
            .map { it.name }
    }
}

interface MyRepo : ReactiveMongoRepository<MyDoc, String> {

}
@Document
class MyDoc(val id:String= UUID.randomUUID().toString(),
    val name:String) {
}